File: PythonDemo/src/demo/Solutions.py
# -*- coding:UTF-8 -*-

'''
Created on 2012-12-18

@author: chengkai
'''

from selenium import webdriver
from selenium.webdriver.support.ui import WebDriverWait 
from selenium.common.exceptions import ElementNotVisibleException
from selenium.webdriver.support.ui import Select
import time
import unittest
import os.path
import os


class SearchPoductsTest(unittest.TestCase):
    

    def setUp(self):
#        self.driver= webdriver.Chrome()
        self.driver= webdriver.Firefox()
        self.driver.get('http://www.360buy.com/')
        print 'preparing this test.'
   
    def tearDown(self):
        self.driver.quit()
        print 'end this test.'

    def testIFrame(self):
        
        WebDriverWait(self.driver, 20, 1, (ElementNotVisibleException)).until
        (lambda driver: driver.find_element_by_id('subscription-submit').is_displayed())
        
        trip_frame = self.driver.find_element_by_xpath("//iframe[@src]")
        self.driver.switch_to_frame(trip_frame)
        
        select_card_price = self.driver.find_element_by_xpath("//*[@id='traveltab']/li[3]")
        print select_card_price.text
        self.driver.switch_to_default_content()
    
    def testMutipleWindow(self):
        WebDriverWait(self.driver, 20, 1, (ElementNotVisibleException)).until
        (lambda driver: driver.find_element_by_id('subscription-submit').is_displayed())
        
        link = self.driver.find_element_by_xpath("//*[@id='jdnews']/div[2]/ul/li[2]/a")
        link.click()
        
        window_handles = self.driver.window_handles
        self.driver.switch_to_window(window_handles[1])
        print self.driver.title
    
    def testDropdown(self):
        WebDriverWait(self.driver, 20, 1, (ElementNotVisibleException)).until
        (lambda driver: driver.find_element_by_id('subscription-submit').is_displayed())
        
        trip_frame = self.driver.find_element_by_xpath("//iframe[@src]")
        self.driver.switch_to_frame(trip_frame)
        self.driver.find_element_by_xpath("//*[@id='traveltab']/li[2]").click()
        dropdown = Select(self.driver.find_element_by_id("hotel-city"))
        options = dropdown.options
        
        for option in options:
            print option.text
        
        self.driver.switch_to_default_content()
    
    def testPageWait(self):
        
        self.driver.implicitly_wait(20)
        trip_frame = self.driver.find_element_by_xpath("//iframe[@src]")
        self.driver.switch_to_frame(trip_frame)
        
        select_card_price = self.driver.find_element_by_xpath("//*[@id='traveltab']/li[3]")
        print select_card_price.text
        self.driver.switch_to_default_content()
        
    def testScreenshot(self):
        WebDriverWait(self.driver, 20, 1, (ElementNotVisibleException)).until
        (lambda driver: driver.find_element_by_id('subscription-submit').is_displayed())
        
        link = self.driver.find_element_by_xpath("//*[@id='jdnews']/div[2]/ul/li[2]/a")
        link.click()
        
        window_handles = self.driver.window_handles
        self.driver.switch_to_window(window_handles[1])
        
        WebDriverWait(self.driver, 20, 1, (ElementNotVisibleException)).until
        (lambda driver: driver.title.startswith('京东商城新闻'))

        print self.driver.title[:6]
        
        try:
            dir = 'c:/img'
            pic_name = '/pic-python.jpg'
            if os.path.isfile(dir+pic_name):
                os.remove(dir+pic_name)
            
            self.driver.get_screenshot_as_file(dir+pic_name)
        except:
            print "fail to get screenshot"
            raise
    
def testSuite():
    suite = unittest.TestSuite()
#    suite.addTest(SearchPoductsTest("testIFrame"))
#    suite.addTest(SearchPoductsTest("testMutipleWindow"))
#    suite.addTest(SearchPoductsTest("testDropdown"))
#    suite.addTest(SearchPoductsTest("testPageWait"))
    suite.addTest(SearchPoductsTest("testScreenshot"))
    return suite
        
        
if __name__ == '__main__':
#    unittest.main()
    suite = testSuite()
    runner = unittest.TextTestRunner()
    runner.run(suite)

File: SeleniumDemo/src/noframe/cases/Solutions.java
package noframe.cases;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Iterator;
import java.util.List;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import org.apache.commons.io.FileUtils;
import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.TakesScreenshot;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.support.ui.ExpectedCondition;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;

import com.thoughtworks.selenium.SeleneseTestBase;

public class Solutions {
	private WebDriver driver = null;
	private SeleneseTestBase testbase = null; 
	
	@Before
	public void setUpBeforeMethod() {
		driver=  new FirefoxDriver();
		testbase = new SeleneseTestBase();
	}
	
//    @Test
	public void assertIframe(){
   	    
        driver.get("http://www.360buy.com/");
        
        (new WebDriverWait(driver, 20)).until(new ExpectedCondition<Boolean>() {
            public Boolean apply(WebDriver d) {
                return d.findElement(By.id("subscription-submit")).isDisplayed();
            }
        });
       
        WebElement trip_frame = driver.findElement(By.xpath("//iframe[@src]")); 
       
        driver.switchTo().frame(trip_frame);
        
        WebElement select_card_price = driver.findElement(By.xpath("//*[@id='traveltab']/li[3]")); 
        System.out.println(select_card_price.getText());
        driver.switchTo().defaultContent();
	}
    
//    @Test
    public void assertPopups(){
   	    
        driver.get("http://www.360buy.com/");
        
        (new WebDriverWait(driver, 20)).until(new ExpectedCondition<Boolean>() {
            public Boolean apply(WebDriver d) {
                return d.findElement(By.id("subscription-submit")).isDisplayed();
            }
        });
       
        WebElement link = driver.findElement(By.xpath("//*[@id='jdnews']/div[2]/ul/li[2]/a")); 
       
        link.click();
        
       Set<String> window_handles = driver.getWindowHandles();
       Iterator<String> ite = window_handles.iterator();
       ite.next();
       driver.switchTo().window(ite.next());
       System.out.println(driver.getTitle());
	}

//    @Test
	public void assertDropdown() {

		driver.get("http://www.360buy.com/");

		(new WebDriverWait(driver, 20)).until(new ExpectedCondition<Boolean>() {
			public Boolean apply(WebDriver d) {
				return d.findElement(By.id("subscription-submit"))
						.isDisplayed();
			}
		});

		WebElement trip_frame = driver.findElement(By.xpath("//iframe[@src]"));

		driver.switchTo().frame(trip_frame);

		driver.findElement(By.xpath("//*[@id='traveltab']/li[2]")).click();
		
		Select city_inhabit = new Select(driver.findElement(By.id("hotel-city")));
		
		
		List<WebElement> options = city_inhabit.getOptions();
		
		for(WebElement element: options){
			System.out.println(element.getText());
		}
		
		driver.switchTo().defaultContent();
	}

//    @Test
	public void assertPageWait() {

		driver.get("http://www.360buy.com/");

		// implicity wait
		driver.manage().timeouts().implicitlyWait(20, TimeUnit.SECONDS);

		WebElement trip_frame = driver.findElement(By.xpath("//iframe[@src]"));

		driver.switchTo().frame(trip_frame);

		driver.findElement(By.xpath("//*[@id='traveltab']/li[2]")).click();
		
		Select city_inhabit = new Select(driver.findElement(By.id("hotel-city")));
		
		List<WebElement> options = city_inhabit.getOptions();
		
		for(WebElement element: options){
			System.out.println(element.getText());
		}
		
		driver.switchTo().defaultContent();
	}
    
    @Test
    public void assertScreenshot() {
    	
         driver.get("http://www.360buy.com/");
        
        (new WebDriverWait(driver, 20)).until(new ExpectedCondition<Boolean>() {
            public Boolean apply(WebDriver d) {
                return d.findElement(By.id("subscription-submit")).isDisplayed();
            }
        });
       
        WebElement link = driver.findElement(By.xpath("//*[@id='jdnews']/div[2]/ul/li[2]/a")); 
       
        link.click();
        
       Set<String> window_handles = driver.getWindowHandles();
       Iterator<String> ite = window_handles.iterator();
       ite.next();
       driver.switchTo().window(ite.next());
       
       (new WebDriverWait(driver, 20)).until(new ExpectedCondition<Boolean>() {
           public Boolean apply(WebDriver d) {
               return d.getTitle().startsWith("京东商城新闻");
           }
       });
       
       File pic = ((TakesScreenshot)driver).getScreenshotAs(OutputType.FILE);
       try {
		FileUtils.copyFile(pic, new File("C:/img/screenshot.jpg"));
	   } catch (IOException e) {
		// TODO Auto-generated catch block
		e.printStackTrace();
	}
       
    }
    
	@After
	public void tearDownAfterMethod() {
		driver.quit();
	}
    
}
